File: datalib/src/main/java/nl/acidcats/tumblrlikes/core/usecases/pincode/PincodeUseCaseImpl.java
package nl.acidcats.tumblrlikes.core.usecases.pincode;

import javax.inject.Inject;

import nl.acidcats.tumblrlikes.core.repositories.AppDataRepository;
import nl.acidcats.tumblrlikes.util.security.SecurityHelper;
import rx.Observable;

/**
 * Created on 28/09/2018.
 */
public class PincodeUseCaseImpl implements PincodeUseCase {

    private AppDataRepository _appDataRepository;
    private SecurityHelper _securityHelper;

    @Inject
    public PincodeUseCaseImpl(AppDataRepository appDataRepository, SecurityHelper securityHelper) {
        _appDataRepository = appDataRepository;
        _securityHelper = securityHelper;
    }

    @Override
    public Observable<Boolean> checkPincode(String pincode) {
        boolean isCorrect;

        if (_appDataRepository.getPincodeHash() != null) {
            String pinCodeHash = _securityHelper.getHash(pincode);
            String storedPinCodeHash = _appDataRepository.getPincodeHash();
            isCorrect = pinCodeHash.equals(storedPinCodeHash);
        } else {
            isCorrect = true;
        }

        return Observable.just(isCorrect);
    }

    @Override
    public Observable<Boolean> storePincode(String pincode) {
        _appDataRepository.setPincodeHash(_securityHelper.getHash(pincode));

        return Observable.just(true);
    }

    @Override
    public Observable<Boolean> isAppPincodeProtected() {
        return Observable.just(_appDataRepository.getPincodeHash() != null);
    }
}

File: datalib/src/main/java/nl/acidcats/tumblrlikes/core/usecases/pincode/PincodeUseCaseImpl.kt
package nl.acidcats.tumblrlikes.core.usecases.pincode

import nl.acidcats.tumblrlikes.core.repositories.AppDataRepository
import nl.acidcats.tumblrlikes.util.security.SecurityHelper
import rx.Observable
import javax.inject.Inject

/**
 * Created on 24/10/2018.
 */
class PincodeUseCaseImpl @Inject constructor(private val appDataRepository: AppDataRepository,
                                             private val securityHelper: SecurityHelper) : PincodeUseCase {

    override fun isAppPincodeProtected(): Observable<Boolean> {
        return Observable.just(appDataRepository.getPincodeHash() != null)
    }

    override fun checkPincode(pincode: String): Observable<Boolean> {
        appDataRepository.getPincodeHash() ?: return Observable.just(true)

        val pinCodeHash = securityHelper.getHash(pincode)
        val storedPinCodeHash = appDataRepository.getPincodeHash()

        return Observable.just(storedPinCodeHash == pinCodeHash)
    }

    override fun storePincode(pincode: String): Observable<Boolean> {
        appDataRepository.setPincodeHash(securityHelper.getHash(pincode))

        return Observable.just(true)
    }
}